scripts/lib/ts-generator.js

```diff
@@ -0,0 +1,437 @@
+ * Custom configuration options are determined by checking the metadata.json file.
+ */
+TSGenerator.prototype.generateCustomConfigFromMetadata = function generateCustomConfigFromMetadata(serviceIdentifier) {
+    // some services have additional configuration options that are defined in the metadata.json file
+    // i.e. dualstackAvailable = useDualstack
+    // create reference to custom options
+    var customConfigurations = [];
+    var serviceMetadata = this.metadata[serviceIdentifier];
+    // loop through metadata members
+    for (var memberName in serviceMetadata) {
+        if (!serviceMetadata.hasOwnProperty(memberName)) {
+            continue;
+        }
+        var memberValue = serviceMetadata[memberName];
+        // check configs
+        switch (memberName) {
+            case 'dualstackAvailable':
+                customConfigurations.push(CUSTOM_CONFIG_ENUMS.DUALSTACK);
+                break;
+        }
+    }
+
+    return customConfigurations;
+};
+
+/**
+ * Generates a type or interface based on the shape.
+ */
+TSGenerator.prototype.generateTypingsFromShape = function generateTypingsFromShape(shapeKey, shape, tabCount) {
+    // some shapes shouldn't be generated if they are javascript primitives
+    if (['string', 'boolean', 'number', 'Date', 'Blob'].indexOf(shapeKey) >= 0) {
+        return '';
+    }
+    var self = this;
+    var code = '';
+    tabCount = tabCount || 0;
+    var tabs = this.tabs;
+    var type = shape.type;
+    if (type === 'structure') {
+        code += tabs(tabCount) + 'export interface ' + shapeKey + ' {\n';
+        var members = shape.members;
+        // cycle through members
+        var memberKeys = Object.keys(members);
+        memberKeys.forEach(function(memberKey) {
+            // docs
+            var member = members[memberKey];
+            if (member.documentation) {
+                code += self.generateDocString(member.documentation, tabCount + 1);
+            }
+            var required = self.checkRequired(shape.required || [], memberKey) ? '?' : '';
+            code += tabs(tabCount + 1) + memberKey + required + ': ' + member.shape + ';\n';
+        });
+        code += tabs(tabCount) + '}\n';
+    } else if (type === 'list') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = ' + shape.member.shape + '[];\n';
+    } else if (type === 'map') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = {[key: string]: ' + shape.value.shape + '};\n';
+    } else if (type === 'string' || type === 'character') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = string;\n';
+    } else if (['double', 'long', 'short', 'biginteger', 'bigdecimal', 'integer', 'float'].indexOf(type) >= 0) {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = number;\n';
+    } else if (type === 'timestamp') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = Date;\n';
```

Comment:
What's an example of a timestamp in the output that is returned as a string, instead of a Date?
If the shape is modeled as a timestamp, and the value is a string or number, we call [`util.date.parseTimestamp`](https://github.com/aws/aws-sdk-js/blob/master/lib/model/shape.js#L269-L274), and [`parseTimestamp`](https://github.com/aws/aws-sdk-js/blob/master/lib/util.js#L291-L305) always returns a Date object, or throws an error.

Fair point about the above for inputs though.